xmtp_mls/src/groups/sync.rs

```diff
@@ -0,0 +1,771 @@
+        if let Err(publish_error) = self.publish_intents(conn).await {
+            log::error!("error publishing intents {:?}", publish_error);
+            errors.push(publish_error);
+        }
+
+        // Even if receiving fails, continue to post_commit
+        if let Err(receive_error) = self.receive(conn).await {
+            log::error!("receive error {:?}", receive_error);
+            // We don't return an error if receive fails, because it's possible this is caused
+            // by malicious data sent over the network, or messages from before the user was
+            // added to the group
+        }
+
+        if let Err(post_commit_err) = self.post_commit(conn).await {
+            log::error!("post commit error {:?}", post_commit_err);
+            errors.push(post_commit_err);
+        }
+
+        // Return a combination of publish and post_commit errors
+        if !errors.is_empty() {
+            return Err(GroupError::Sync(errors));
+        }
+
+        Ok(())
+    }
+
+    /**
+     * Sync the group and wait for the intent to be deleted
+     * Group syncing may involve picking up messages unrelated to the intent, so simply checking for errors
+     * does not give a clear signal as to whether the intent was successfully completed or not.
+     *
+     * This method will retry up to `crate::configuration::MAX_GROUP_SYNC_RETRIES` times.
+     */
+    pub(super) async fn sync_until_intent_resolved<'a>(
+        &self,
+        conn: &'a DbConnection<'a>,
+        intent_id: ID,
+    ) -> Result<(), GroupError> {
+        let mut num_attempts = 0;
+        // Return the last error to the caller if we fail to sync
+        let mut last_err: Option<GroupError> = None;
+        while num_attempts < crate::configuration::MAX_GROUP_SYNC_RETRIES {
+            if let Err(err) = self.sync_with_conn(conn).await {
+                log::error!("error syncing group {:?}", err);
+                last_err = Some(err);
+            }
+
+            // This will return early if the fetch fails
+            let intent: Result<Option<StoredGroupIntent>, StorageError> = conn.fetch(&intent_id);
+            match intent {
+                Ok(None) => {
+                    // This is expected. The intent gets deleted on success
+                    return Ok(());
+                }
+                Ok(Some(intent)) => {
+                    log::warn!(
+                        "retrying intent ID {}. intent currently in state {:?}",
+                        intent.id,
+                        intent.state
+                    );
+                }
+                Err(err) => {
+                    log::error!("database error fetching intent {:?}", err);
+                    last_err = Some(GroupError::Storage(err));
+                }
+            };
+            num_attempts += 1;
+        }
+
+        Err(last_err.unwrap_or(GroupError::Generic("failed to wait for intent".to_string())))
+    }
+
+    fn process_own_message(
```

Comment:
Everything below this has been moved over from `groups/mod.rs` unless otherwise noted